internal/dag/gatewayapi_processor.go

```diff
@@ -1119,9 +1119,47 @@ func (p *GatewayAPIProcessor) computeHTTPRoute(route *gatewayapi_v1beta1.HTTPRou
 				// Get the redirect filter if there is one. Note that per Gateway API
 				// docs, "specifying a core filter multiple times has unspecified or
 				// custom conformance.", here we choose to just select the first one.
+				if pathRewritePolicy != nil {
+					continue
+				}
+
 				if redirect == nil && filter.RequestRedirect != nil {
 					redirect = filter.RequestRedirect
 				}
+
+				if redirect.Path == nil {
+					continue
+				}
+
+				var prefixRewrite, fullPathRewrite string
+
+				switch redirect.Path.Type {
+				case gatewayapi_v1beta1.PrefixMatchHTTPPathModifier:
+					if filter.RequestRedirect.Path.ReplacePrefixMatch == nil || len(*filter.RequestRedirect.Path.ReplacePrefixMatch) == 0 {
+						prefixRewrite = "/"
+					} else {
+						prefixRewrite = *filter.RequestRedirect.Path.ReplacePrefixMatch
+					}
+				case gatewayapi_v1beta1.FullPathHTTPPathModifier:
+					if filter.RequestRedirect.Path.ReplaceFullPath == nil || len(*filter.RequestRedirect.Path.ReplaceFullPath) == 0 {
+						fullPathRewrite = "/"
+					} else {
+						fullPathRewrite = *filter.RequestRedirect.Path.ReplaceFullPath
+					}
+				default:
+					routeAccessor.AddCondition(
+						gatewayapi_v1beta1.RouteConditionAccepted,
+						metav1.ConditionFalse,
+						gatewayapi_v1beta1.RouteReasonUnsupportedValue,
+						fmt.Sprintf("HTTPRoute.Spec.Rules.Filters.RequestRedirect.Path.Type: invalid type %q: only ReplacePrefixMatch and ReplaceFullPath are supported.", filter.RequestRedirect.Path.Type),
+					)
+					continue
+				}
+
+				pathRewritePolicy = &PathRewritePolicy{
+					PrefixRewrite:   prefixRewrite,
+					FullPathRewrite: fullPathRewrite,
+				}
```

Comment:
We should not be using `pathRewritePolicy` here since that gets applied for a Route that proxies to an upstream; instead, these rewrite fields need to be used for constructing the Redirect action which is happening [here](https://github.com/projectcontour/contour/blob/ed3dbafa3e372913bf5d5325e63012c9bbc73ed2/internal/dag/gatewayapi_processor.go#L1630-L1635) for the DAG and then [here](https://github.com/projectcontour/contour/blob/ed3dbafa3e372913bf5d5325e63012c9bbc73ed2/internal/envoy/v3/route.go#L242) to generate Envoy config.

What I would probably suggest is moving the instantiation of the `dag.Redirect` that is currently happening [here](https://github.com/projectcontour/contour/blob/ed3dbafa3e372913bf5d5325e63012c9bbc73ed2/internal/dag/gatewayapi_processor.go#L1630-L1635) up to here, and setting the appropriate rewrite fields on it as part of that. Then, pass that `dag.Redirect` into `p.redirectRoutes` instead of the `gatewayapi_v1beta1.HTTPRequestRedirectFilter`.

Note that `dag.Redirect` and the `routeRedirect` function both already have support for prefix and full path rewriting since we [already support this for HTTPProxy](https://projectcontour.io/docs/v1.23.2/config/api/#projectcontour.io/v1.HTTPRequestRedirectPolicy). However, I think it's possible there's a bug in that implementation that doesn't handle completely removing the path, i.e. replacing it with `/`. You may need to use similar logic as https://github.com/projectcontour/contour/blob/ed3dbafa3e372913bf5d5325e63012c9bbc73ed2/internal/dag/gatewayapi_processor.go#L1551-L1568 to make this work for rewrites during redirects.